app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/MockNavigationOffboardRouterActivity.kt

```diff
@@ -0,0 +1,372 @@
+            locationEngine = ReplayRouteLocationEngine()
+            newOrigin()
+        }
+    }
+
+    override fun onMapClick(point: LatLng): Boolean {
+        when {
+            destination == null -> {
+                destination = Point.fromLngLat(point.longitude, point.latitude)
+                mapboxMap?.addMarker(MarkerOptions().position(point))
+            }
+            waypoint == null -> {
+                waypoint = Point.fromLngLat(point.longitude, point.latitude)
+                mapboxMap?.addMarker(MarkerOptions().position(point))
+            }
+            else -> Toast.makeText(this, "Only 2 waypoints supported", Toast.LENGTH_LONG).show()
+        }
+        calculateRoute()
+        return false
+    }
+
+    private fun onStartRouteClick() {
+        ifNonNull(route, locationEngine) { route, locationEngine ->
+            val isValidRoute = route.distance()?.let { it > TWENTY_FIVE_METERS } ?: false
+            if (isValidRoute) {
+                // Hide the start button
+                startRouteButton.visibility = View.INVISIBLE
+
+                // Attach all of our navigation listeners.
+                navigation.addNavigationEventListener(this)
+                navigation.addProgressChangeListener(this)
+                navigation.addMilestoneEventListener(this)
+                navigation.addOffRouteListener(this)
+
+                (locationEngine as ReplayRouteLocationEngine).assign(route)
+                navigation.locationEngine = locationEngine
+                mapboxMap?.let { it.locationComponent.isLocationComponentEnabled = true }
+                navigation.startNavigation(route)
+                mapboxMap?.removeOnMapClickListener(this)
+            }
+        }
+    }
+
+    private fun onNewLocationClick() {
+        newOrigin()
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun calculateRoute() {
+        locationEngine.getLastLocation(object : LocationEngineCallback<LocationEngineResult> {
+            override fun onSuccess(result: LocationEngineResult) {
+                findRouteWith(result)
+            }
+
+            override fun onFailure(exception: Exception) {
+                Timber.e(exception)
+            }
+        })
+    }
+
+    private fun findRouteWith(result: LocationEngineResult) {
+        val userLocation = result.lastLocation
+        if (userLocation == null) {
+            Timber.d("calculateRoute: User location is null, therefore, origin can't be set.")
+            return
+        }
+        val origin = Point.fromLngLat(userLocation.longitude, userLocation.latitude)
+        destination?.let { destinationPoint ->
+            if (TurfMeasurement.distance(origin, destinationPoint, TurfConstants.UNIT_METERS) < 50) {
+                startRouteButton.visibility = View.GONE
+                return
+            }
+
+            val waypoints = mutableListOf<Point>()
```

Comment:
```suggestion
            val waypoints = mutableListOf(waypoint).mapNotNull()
```

Reply:
You probably mean `filterNotNull()`) But yeah, anyway, you are right